docs/development/design/watch-api.md

```diff
@@ -0,0 +1,191 @@
+
+A Informer in essence is a local cached copy of the resources that clients are interested in. The local cache will be refreshed via the watch operation. But the use of the watch API is occurring asynchronously and hidden from the clients.
+
+To use the Informer API
+
+```go
+informer := client.DevicesApi.
+   ListDevicesInOrganization(context.Background(), orgID).
+   Informer()
+
+devices, _, err := informer.Execute()
+```
+
+Notice that result of `informer.Execute()` is the same as the traditional list style api.  Additionally an informer provides a chanel via `informer.Changed()` that you can use to wait for when you can call `informer.Execute()` to get a changed device list.
+
+```go
+informer := client.DevicesApi.
+   ListDevicesInOrganization(ctx, orgID).
+   Informer()
+
+for {
+    select {
+        case <- ctx.Done():
+            return
+        case <- informer.Changed():
+            devices, _, err := informer.Execute()
+            ... process the devices ....
+    }
+}
+```
+
+### Apiserver Implementation of `watch=true`
+
+The HTTP request handler servicing the `ListDevicesInOrganization` will:
+
+1. Create a **SignalBus** (described later in this do) subscription to `/devices/org={orgId}`
+2. Select matching devices from the DB.  It will force device results to be ordered by `revision` so that the last selected result has the highest revision.
+3. Each result will be sent to the client as independent JSON document.  This will continue until it gets an empty result set from the database.
+4. At that point it will send the bookmark event the the clien
+5. Parks the go routine waiting for the http requent to be terminated or the **SignalBus** subcription to be notified.
+6. It loops back and selects matching devices from the DB since that last seen revision.
+7. Sends each selected device to the client,
+8. loops back to #5
+
+All API operations that change the desired device state singal the **SignalBus** on the `/devices/org={orgId}` when those devices are changed.
+
+### What is a SignalBus
+
+A **SignalBus** is a really simple interface that you can use to notify other go routines that a named thing has been signaled.
+
+```go
+type SignalBus interface {
+    // Notify will notify all the subscriptions created for the given named signal.
+    Notify(name string)
+    // NotifyAll will notify all the subscriptions
+    NotifyAll()
+    // Subscribe creates a subscription the named signal
+    Subscribe(name string) *Subscription
+}
+
+type Subscription interface { 
+    // You can wait on this channel to know when it's been signaled
+    Signal() <-chan bool
+    // close out the subscription.
+    Close()
+}
+```
+
+Since this interface does not hold events and even coalesces multiple Notify calls.  It results in always being non-blocking to the callers of Notify and always have a bounded amount of memory that it uses.
```

Comment:
```suggestion
Since this interface does not hold events and even coalesces multiple Notify calls, it results in always being non-blocking to the callers of Notify and always having a bounded amount of memory that it uses.
```